core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -355,71 +475,90 @@ int findVersion() {
-        RuntimeException re = tryDelete(src);
-        if (re != null) {
-          cfe.addSuppressed(re);
-        }
-        throw cfe;
+      if (fs.exists(finalMetaDataFile)) {
+        throw new CommitFailedException(
+            "Version %d already exists: %s", nextVersion, finalMetaDataFile);
       }
-    } catch (IOException e) {
-      CommitFailedException cfe =
-          new CommitFailedException(e, "Failed to commit changes using rename: %s", dst);
-      RuntimeException re = tryDelete(src);
-      if (re != null) {
-        cfe.addSuppressed(re);
+      // maybe too heavy.....?
+      if (!nextVersionIsLatest(nextVersion, fs)) {
+        // In the case of concurrent execution,
+        // verify that the version that is ready to be committed at a time is the latest version.
+        throw new CommitFailedException("Version %d too old: %s", nextVersion, finalMetaDataFile);
       }
-      throw cfe;
+      return renameMetaDataFileAndCheck(fs, tempMetaDataFile, finalMetaDataFile);
     } finally {
-      if (!lockManager.release(dst.toString(), src.toString())) {
-        LOG.warn("Failed to release lock on file: {} with owner: {}", dst, src);
+      if (!lockManager.release(finalMetaDataFile.toString(), finalMetaDataFile.toString())) {
+        LOG.warn(
+            "Failed to release lock on file: {} with owner: {}",
+            finalMetaDataFile,
+            tempMetaDataFile);
       }
     }
   }
 
-  /**
-   * Deletes the file from the file system. Any RuntimeException will be caught and returned.
-   *
-   * @param path the file to be deleted.
-   * @return RuntimeException caught, if any. null otherwise.
-   */
-  private RuntimeException tryDelete(Path path) {
-    try {
-      io().deleteFile(path.toString());
-      return null;
-    } catch (RuntimeException re) {
-      return re;
-    }
+  private void cleanUncommittedMeta(Path src) {
+    io().deleteFile(src.toString());
   }
 
   protected FileSystem getFileSystem(Path path, Configuration hadoopConf) {
     return Util.getFs(path, hadoopConf);
   }
 
+  @VisibleForTesting
+  boolean renameMetaDataFileAndCheck(FileSystem fs, Path tempMetaDataFile, Path finalMetaDataFile) {
+    try {
+      // The most important step. There must be no mistakes in this step.
+      // Even if it does, we should stop everything.
+      return renameMetaDataFile(fs, tempMetaDataFile, finalMetaDataFile);
+    } catch (Throwable e) {
+      LOG.error("There were some problems with submitting the new version.", e);
+      try {
+        if (newMetadataExists(fs, finalMetaDataFile) && !tempMetadataExists(fs, tempMetaDataFile)) {
+          return true;
+        } else {
+          throw new CommitFailedException(e);
+        }
+      } catch (CommitFailedException e1) {
+        throw e1;
+      } catch (Throwable e2) {
+        throw new CommitStateUnknownException(e2);
+      }
+    }
+  }
+
+  @VisibleForTesting
```

Comment:
Why are there two methods here? They both do exactly the same thing and i'm not sure they are any more succint then just inlining fs.exists

Reply:
With OOP programming you don't want to assert the calls, but want to validate the behavior of the HadoopTableOperations. In order to do this you would create scenario's where you use Mockito to let certain operations fail (for example `fs.rename`), and then you assert the desired state. Opening up the object for testing is in general considered a bad practice.

An example can be found here: https://github.com/apache/iceberg/blob/29fd2a0cb940fd16aa5bfb8003ec1354e678c949/core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java#L329-L343

Reply:
@Fokko 
Yes, but in practice many detailed scenarios are difficult to create locally. ...... It is difficult to simulate such scenarios. For example: halfway through a method call, simulating all accesses to the filesystem suddenly throws an exception. I don't know how to simulate this scenario.